[2단계 - 점진적인 리팩터링] 도기(이동엽) 미션 제출합니다.

dye0p opened the pull request:
안녕하세요 노랑~ 2단계 미션 제출합니다! 
이번 미션은 `HandlerAdapter` + `HandlerMapping` + `DispatcherServlet`가 역할, 책임을 잘 수행할 수 있도록 구현하는 것을 목표로 했습니다! 각 객체들의 역할, 책임과 협력이 잘 이루어지고 있는지 중점적으로 봐주시면 좋을 것 같아요!
리뷰 잘 부탁드립니다🙇‍♂️

dye0p committed:
refactor: 컨트롤러 스캔 로직 분리 및 AnnotationHandlerMapping 개선

- ControllerScanner 클래스 추가로 컨트롤러 스캔 책임 분리
- Reflections 의존성을 활용한 getControllers 메서드 구현
- AnnotationHandlerMapping에서 scanControllers 제거 및 초기화 로직 간소화
- 중복된 컨트롤러 인스턴스 생성 로직 제거

dye0p committed:
refactor: move 메서드명 render로 수정

- 메서드명을 render로 변경하여 역할 명확화
- DispatcherServlet에서 호출되는 메서드 이름 일관성 확보

dye0p committed:
feat: HandlerMapping, HandlerAdapter 인터페이스 추가

dye0p committed:
feat: 다양한 핸들러를 지원하는 HandlerAdapter 구현체 추가

dye0p committed:
refactor: 기존 HandlerMapping 구현체에 HandlerMapping 인터페이스 적용

dye0p committed:
refactor: 기존 Controller가 ModelAndView를 반환하도록 수정

dye0p committed:
refactor: DispatcherServlet이 HandlerMapping과 HandlerAdapter를 사용하도록 리팩터링

dye0p committed:
refactor: 상수 이름 가독성 향상

- `INTERFACE_21`을 `basePackage`로 변경하여 의미 명확화
- 관련 초기화 코드에서 변경된 상수명 적용

yesjuhee commented on mvc/src/main/java/com/interface21/webmvc/servlet/mvc/tobe/ControllerScanner.java:
```diff
@@ -0,0 +1,40 @@
+            throw new RuntimeException(e);
+        }
+    }
+
```
이 개행은 삭제해도 될 것 같습니다!

dye0p replied:
제거 하였습니다!

[refactor: 불필요한 개행 제거](https://github.com/dye0p/java-mvc/commit/feb7e7efba719dd4966825c39244816b37d3d699)

yesjuhee commented on mvc/src/main/java/com/interface21/webmvc/servlet/mvc/tobe/ControllerScanner.java:
```diff
@@ -0,0 +1,40 @@
+    }
+
+    private Map<Class<?>, Object> instantiateControllers(final Set<Class<?>> controllersClasses) {
+        return controllersClasses.stream()
+                .collect(Collectors.toMap(
+                        controllerClass -> controllerClass,
+                        this::getInstanceBy
+                ));
```
```suggestion
        return controllersClasses.stream()
                .collect(Collectors.toMap(
                        Function.identity(),
                        this::getInstanceBy
                ));
```

자기 자신을 반환하는 함수의 경우 `Function.identity()`를 활용할 수 있습니다!

참고: https://juno-juno.tistory.com/103

dye0p replied:
자료 첨부까지! 감사합니다 :)

[refactor: Function.identity()로 가독성 개선](https://github.com/dye0p/java-mvc/commit/f59d4498e39d61a1369a6f9ed5f9cc7c56aee53c)

yesjuhee commented on mvc/src/main/java/com/interface21/webmvc/servlet/mvc/tobe/AnnotationHandlerMapping.java:
```diff
@@ -27,13 +24,20 @@ public AnnotationHandlerMapping(final Object... basePackage) {
     }
 
+    @Override
     public Object getHandler(final HttpServletRequest request) {
         final String url = request.getRequestURI();
```
```suggestion
    public HandlerExecution getHandler(final HttpServletRequest request) {
        final String url = request.getRequestURI();
```

타입 안정성과 가독성을 위해 반환 타입을 더 구체적으로 명시하는거 어떨까요?

dye0p replied:
그게 좋겠네요! 수정했습니다👍 

[refactor: AnnotationHandlerMapping getHandler 반환 타입 변경](https://github.com/dye0p/java-mvc/commit/76bc2c50c912ce74f7cf1bf34833bcce760cefc5)

yesjuhee commented on mvc/src/main/java/com/interface21/webmvc/servlet/mvc/asis/Controller.java:
```diff
@@ -1,8 +1,9 @@
 import jakarta.servlet.http.HttpServletResponse;
 
 public interface Controller {
-    String execute(final HttpServletRequest req, final HttpServletResponse res) throws Exception;
+    ModelAndView execute(final HttpServletRequest req, final HttpServletResponse res) throws Exception;
```
`Controller` 인터페이스가 변경되어서 어플리케이션이 실행되지 않습니다!!

`Controller` 인터페이스를 구현하고 있는 `app/src/main/java/com/techcourse/controller` 패키지들이 컴파일 되지 않기 때문에 어플리케이션 실행이 되지 않고 있어요!

---

### execute 는 String 반환으로 유지

이를 해결하기 위해 `controller` 패티지 하위의 컨트롤러들이 `ModelAndView`를 반환하도록 수정해서 해결할 수도 있습니다. 

하지만 미션의 의도에 맞게 해결하려면 **`Controller` 인터페이스가 `String`을 반환하도록 유지**한 상황에서 `ManuHandlerMapping`과 `AnnotationHandlerMapping`이 둘 다 작동되도록 수정해야합니다.

LMS을 보면 다음과 같은 미션 설명이 있습니다.
> 새로운 MVC 프레임워크를 추가하면 기존에 구현한 컨트롤러 인터페이스 기반 MVC 프레임워크로 만든 컨트롤러도 변경 해야 할까?

지금과 같이 Controlelr의 반환 타입을 String 에서 ModelAndView로 바꾸는 것이, 새로운 MVC 프레임워크를 추가함으로써 기존에 구현한 컨트롤러 인터페이스 기반 MVC 프레임워크로 만든 컨트롤러 전체를 변경하게 되는 상황에 해당한다고 생각합니다.

따라서 **기존의 Controller 인터페이스를 변경하지 않고** 두가지 HandelrMapping이 모두 동작할 수 있도록 코드를 작성하는게 이번 미션의 목표입니다!

---

혹시 잘 이해가 안되거나 궁금한 점이 있으시면 언제든지 DM or 찾아와서 질문 주세욥

dye0p replied:
노랑의 의견을 듣고보니 맞는 것 같아요! 제가 요구사항을 제대로 이해하지 못한 것 같습니다! 말씀하신대로 기존의 코드 변경없이 동작할 수 있도록 수정하였습니다! 혹시 부족한 부분이 있다면 말씀해주세요! 좋은 리뷰 감사합니다 :)

[refactor: ManualHandlerAdapter가 View와 ModelAndView 객체를 생성하도록 수정](https://github.com/dye0p/java-mvc/commit/58cabb622c8b2e504b2501980a9b5b91d8b7ccff)
[refactor: Controller 인터페이스 execute 메서드 반환 타입 변경](https://github.com/dye0p/java-mvc/commit/1320096f635addb6f0a98b8d5af0e71cd757e354)

yesjuhee reviewed (changes requested):
안녕하세요 도기!!
다양한 객체들이 서로 협력해서 일을 잘 수행하고 있는 코드 확인했습니다 👍 
다만 컨트롤러 인터페이스 변경으로 Application 실행이 되지 않고 있어서 이부분 리뷰 남겼으니 확인 부탁드립니다 😄

dye0p committed:
refactor: 불필요한 개행 제거

dye0p committed:
refactor: Function.identity()로 가독성 개선

- Collectors.toMap에서 Function.identity() 사용으로 람다 표현식 단축
- 코드 간결성과 가독성 향상

dye0p committed:
refactor: Controller 인터페이스 execute 메서드 반환 타입 변경

- 기존의 형태를 유지하기 위해서 String으로 변환

dye0p committed:
refactor: ManualHandlerAdapter가 View와 ModelAndView 객체를 생성하도록 수정

- Controller의 execute 메서드 반환값(String)을 기반으로 JspView 생성
- JspView를 사용해 ModelAndView 객체 반환하도록 변경

dye0p committed:
refactor: AnnotationHandlerMapping getHandler 반환 타입 변경

- getHandler 메서드 반환 타입을 Object에서 HandlerExecution으로 변경
- 코드 명확성과 타입 체크 개선

dye0p committed:
test: 불필요한 타입 캐스팅 제거

dye0p reviewed (commented):
안녕하세요 노랑! 리뷰 반영후 다시 요청드려요!  혹시 피드백이 있다면 남겨주세요~

yesjuhee commented on app/src/main/java/com/techcourse/DispatcherServlet.java:
```diff
@@ -1,28 +1,39 @@
 public class DispatcherServlet extends HttpServlet {
 
     private static final long serialVersionUID = 1L;
+    private static final String basePackage = "interface21";
```
```suggestion
    private static final String basePackage = "techcourse.controller";
```

basePackage 의 설정 값이 수정되어야 합니다!

어노테이션 컨트롤러를 작성한다면, `app/src/main/java/com/techcourse/controller` 하위에 작성하게 될 것 같은데요. 지금 코드에서는 basePackage가 mvc 프레임워크 영역에 해당하는 `interface21`로 설정되어 있어서 **`app/` 하위의 controller를 찾을 수 없습니다.**

이 부분을 수정 한 다음에 `com.techcourse.controller`에 아래 테스트 컨트롤러 추가 후, `http://localhost:8080/test` 위치로 접근해서 `index.jsp`를 볼 수 있는지 확인해보세요!

```java
@Controller
public class Step2Controller {

    @RequestMapping(value = "/test", method = RequestMethod.GET)
    public ModelAndView showForm(HttpServletRequest req, HttpServletResponse res) {
        final var modelAndView = new ModelAndView(new JspView("/index.jsp"));
        return modelAndView;
    }
}
```

dye0p replied:
스캔 범위 지정을 했어야 했는데 못했네요!! 꼼꼼하게 찾아주셔서 감사해요 노랑! 테스트 컨트롤러까지 같이 주셔서 감사해요. 수정했습니다!

yesjuhee reviewed (changes requested):
안녕하세요 도기 😄 
두 가지 방식의 컨트롤러가 모두 동작하려면 하나의 설정이 더 필요합니다!
해당 부분까지 완료하고 step2 마무리하는게 좋을 것 같아서 RC로 보내드립니다~!

(참고로 저도 아직 step3 시작 안해서 천천히 하셔도 괜찮아요 🙌 )

dye0p committed:
fix: 어노테이션 기반의 Controller를 HandlerMapping이 찾을 수 있도록 basePackage 경로 수정

- TestController를 추가하여 동작 확인
